File: packages/core/src/http/raw-request.ts
import type { HttpRequestShape, RawHttpRequestParseOptions } from '../types/public.js'
import { normalizeHeaderName, normalizeHeaderValue } from './headers.js'

const REQUEST_LINE_PATTERN = /^(\S+)\s+(\S+)(?:\s+HTTP\/\d(?:\.\d)?)?$/

const resolveRawRequestUrl = (
  target: string,
  headers: Record<string, string>,
  options: RawHttpRequestParseOptions
): string => {
  if (/^https?:\/\//i.test(target)) {
    return new URL(target).toString()
  }

  const authority = headers.host ?? headers[':authority']

  if (!authority) {
    throw new Error(
      'A raw HTTP request with a relative request target must include a Host header or use an absolute target URL.'
    )
  }

  const defaultScheme = options.defaultScheme?.trim() || 'https'

  return new URL(target, `${defaultScheme}://${authority}`).toString()
}

/**
 * Parses a captured HTTP request string into the shared request model used by the toolkit.
 */
export const parseRawHttpRequest = (
  rawRequest: string,
  options: RawHttpRequestParseOptions = {}
): HttpRequestShape => {
  const normalizedInput = rawRequest.replace(/\r\n/g, '\n').trim()

  if (!normalizedInput) {
    throw new Error('A raw HTTP request is required.')
  }

  const separatorIndex = normalizedInput.indexOf('\n\n')
  const head = separatorIndex === -1 ? normalizedInput : normalizedInput.slice(0, separatorIndex)
  const body = separatorIndex === -1 ? '' : normalizedInput.slice(separatorIndex + 2)
  const lines = head.split('\n')
  const requestLine = lines.shift()?.trim()

  if (!requestLine) {
    throw new Error('The raw HTTP request is missing a request line.')
  }

  const requestLineMatch = REQUEST_LINE_PATTERN.exec(requestLine)

  if (!requestLineMatch) {
    throw new Error(
      'The raw HTTP request line must be formatted as "METHOD /path HTTP/1.1" or use an absolute target URL.'
    )
  }

  const method = requestLineMatch[1]
  const target = requestLineMatch[2]

  if (!method || !target) {
    throw new Error(
      'The raw HTTP request line must include both an HTTP method and a request target.'
    )
  }

  const headers: Record<string, string> = {}
  let currentHeaderName: string | undefined

  for (const rawLine of lines) {
    if (!rawLine.trim()) {
      continue
    }

    if (/^\s/.test(rawLine) && currentHeaderName) {
      headers[currentHeaderName] = `${headers[currentHeaderName]} ${rawLine.trim()}`
      continue
    }

    const separatorIndex = rawLine.indexOf(':')

    if (separatorIndex === -1) {
      throw new Error(`The raw HTTP header line "${rawLine}" must be formatted as "Name: value".`)
    }

    const name = normalizeHeaderName(rawLine.slice(0, separatorIndex))
    const value = normalizeHeaderValue(rawLine.slice(separatorIndex + 1))

    if (!name || !value) {
      throw new Error(`The raw HTTP header line "${rawLine}" must include both a name and a value.`)
    }

    headers[name] = headers[name] ? `${headers[name]}, ${value}` : value
    currentHeaderName = name
  }

  return {
    ...(body ? { body } : {}),
    headers,
    method,
    url: resolveRawRequestUrl(target, headers, options)
  }
}

File: languages/python/src/open_payments_http_signatures_devkit/raw_request.py
"""Raw HTTP request parsing for captured traces.

Parses a raw HTTP request string (as captured from tools like curl --trace,
tcpdump, or proxy logs) into the shared HttpRequest model used by the toolkit.
"""

from __future__ import annotations

import re
from urllib.parse import urljoin

from .request import normalize_request
from .types import HttpRequest
from .utils import normalize_headers

_REQUEST_LINE_RE = re.compile(r"^(\S+)\s+(\S+)(?:\s+HTTP/\d(?:\.\d)?)?$")


def _resolve_raw_request_url(
    target: str,
    headers: dict[str, str],
    default_scheme: str,
) -> str:
    if re.match(r"^https?://", target, re.IGNORECASE):
        return target

    authority = headers.get("host") or headers.get(":authority")

    if not authority:
        raise ValueError(
            "A raw HTTP request with a relative request target must include "
            "a Host header or use an absolute target URL."
        )

    base = f"{default_scheme}://{authority}"
    return urljoin(base, target)


def parse_raw_http_request(
    raw_request: str,
    *,
    default_scheme: str = "https",
) -> HttpRequest:
    """Parse a captured raw HTTP request string into an HttpRequest.

    Supports both ``\\r\\n`` and ``\\n`` line endings. A blank line separates
    headers from the optional body.  Continuation lines (starting with
    whitespace) are folded into the preceding header value.

    Args:
        raw_request: The raw HTTP request text.
        default_scheme: Scheme to use when the request target is relative
            and a Host header is present.  Defaults to ``"https"``.

    Returns:
        An :class:`HttpRequest` with the parsed method, URL, headers, and body.

    Raises:
        ValueError: If the input is empty or the request line is malformed.
    """
    normalized_input = raw_request.replace("\r\n", "\n").strip()

    if not normalized_input:
        raise ValueError("A raw HTTP request is required.")

    separator_index = normalized_input.find("\n\n")

    if separator_index == -1:
        head = normalized_input
        body = ""
    else:
        head = normalized_input[:separator_index]
        body = normalized_input[separator_index + 2 :]

    lines = head.split("\n")
    request_line = lines.pop(0).strip() if lines else ""

    if not request_line:
        raise ValueError("The raw HTTP request is missing a request line.")

    match = _REQUEST_LINE_RE.match(request_line)

    if not match:
        raise ValueError(
            'The raw HTTP request line must be formatted as '
            '"METHOD /path HTTP/1.1" or use an absolute target URL.'
        )

    method = match.group(1)
    target = match.group(2)

    if not method or not target:
        raise ValueError(
            "The raw HTTP request line must include both an HTTP method "
            "and a request target."
        )

    headers: dict[str, str] = {}
    current_header_name: str | None = None

    for raw_line in lines:
        if not raw_line.strip():
            continue

        # Continuation line (obs-fold)
        if raw_line[0] in (" ", "\t") and current_header_name:
            headers[current_header_name] = (
                f"{headers[current_header_name]} {raw_line.strip()}"
            )
            continue

        colon_index = raw_line.find(":")

        if colon_index == -1:
            raise ValueError(
                f'The raw HTTP header line "{raw_line}" must be formatted '
                'as "Name: value".'
            )

        name = raw_line[:colon_index].strip().lower()
        value = raw_line[colon_index + 1 :].strip()

        if not name or not value:
            raise ValueError(
                f'The raw HTTP header line "{raw_line}" must include '
                "both a name and a value."
            )

        if name in headers:
            headers[name] = f"{headers[name]}, {value}"
        else:
            headers[name] = value

        current_header_name = name

    url = _resolve_raw_request_url(target, headers, default_scheme)

    result = HttpRequest(
        method=method,
        url=url,
        headers=headers if headers else None,
        body=body if body else None,
    )

    return result
